args/core/scheduling/scheduler.cpp

```diff
@@ -6,22 +6,90 @@ namespace args::core::scheduling
 {
     async::readonly_rw_spinlock Scheduler::m_threadsLock;
     sparse_map<std::thread::id, std::unique_ptr<std::thread>> Scheduler::m_threads;
-    const uint Scheduler::m_maxThreadCount = std::thread::hardware_concurrency() == 0 ? std::numeric_limits<uint>::max() : std::thread::hardware_concurrency();
+    std::queue<std::thread::id> Scheduler::m_unreservedThreads;
+    const uint Scheduler::m_maxThreadCount = (((int)std::thread::hardware_concurrency()) - 2) <= 0 ? 4 : std::thread::hardware_concurrency();
     async::readonly_rw_spinlock Scheduler::m_availabilityLock;
     uint Scheduler::m_availableThreads = m_maxThreadCount - 2; // subtract OS and this_thread.
 
-    Scheduler::Scheduler(events::EventBus* eventBus) : m_eventBus(eventBus)
+    async::readonly_rw_spinlock Scheduler::m_jobQueueLock;
+    std::queue<Scheduler::runnable> Scheduler::m_jobs;
+    sparse_map<std::thread::id, async::readonly_rw_spinlock> Scheduler::m_commandLocks;
+    sparse_map<std::thread::id, std::queue<Scheduler::runnable>> Scheduler::m_commands;
+
+    void Scheduler::threadMain(bool* exit, bool* start, bool low_power)
+    {
+        std::thread::id id = std::this_thread::get_id();
+
+        while (!(*start))
+        {
+            if (low_power)
+                std::this_thread::sleep_for(std::chrono::milliseconds(1));
+            else
+                std::this_thread::yield();
+        }
+
+        while (!(*exit))
+        {
+            Scheduler::runnable instruction{};
+
+            {
+                async::readonly_guard guard(m_commandLocks[id]);
+                if (!m_commands[id].empty())
+                {
+                    instruction = m_commands[id].front();
+                    async::readwrite_guard rwguard(m_commandLocks[id]);
+                    m_commands[id].pop();
```

Comment:
isn't this a potential bug ?
consider this scenario:
A & B both lock for reading
A gets front (aka has a command)
B gets front (aka has the same command)

_now_ A locks for deleting with pop()
**B now either fails deleting or can never aquire the lock**

Scenario 1 (B fails deleting) 
 - B triggers an exception (depending on how pop works) or
 - B & A execute the instruction twice

Scenario 2 (B can not aquire the lock)
 - B effectively halts.

Reply:
potential work around is to check if the job still exists after aquiring the read-lock